distributed/tests/test_scheduler.py

```diff
@@ -3538,3 +3541,84 @@ async def test_repr(s, a):
             repr(ws_b)
             == f"<WorkerState {b.address!r}, status: running, memory: 0, processing: 0>"
         )
+
+
+@gen_cluster(client=True)
+async def test_stimulus_success(c, s, a, b):
+    f = c.submit(inc, 1)
+    key = f.key
+
+    await f
+    await c.close()
+
+    stories = s.story(key)
+
+    assert_story(
+        stories,
+        [
+            (key, "released", "waiting", {key: "processing"}),
+            (key, "waiting", "processing", {}),
+            (key, "processing", "memory", {}),
+            (
+                key,
+                "memory",
+                "forgotten",
+                {},
+            ),
+        ],
+    )
+
+    stimulus_ids = {s[-2] for s in stories}
+    assert len(stimulus_ids) == 3
+
+
+@gen_cluster(client=True)
+async def test_stimulus_retry(c, s, a, b):
+    def task():
+        assert dask.config.get("foo")
+
+    with dask.config.set(foo=False):
+        f = c.submit(task)
+        with pytest.raises(AssertionError):
+            await f
+
+    with dask.config.set(foo=True):
+        await f.retry()
+        await f
+
+    story = s.story(f.key)
+    stimulus_ids = {s[-2] for s in story}
+    assert len(stimulus_ids) == 4
```

Comment:
Why four exactly?  If a future version of Dask made this 5 or 3 would that be incorrect?  I think instead you're looking for a statement like "there are far fewer stimuli than there are events in the log" .  Maybe `assert len(stimulus_ids) < len(story) / 3` ?

Reply:
My understanding of yesterday's conversation with @fjetter was that testing the number of unique stimulus_id's was the strategy to use at present.

Reply:
Right, but we don't need to be exact.  What if a future change in the code change this number so that it became 3?  Would that be a sign of an error?  What if it was 5?  

A change of that sort is likely to occur in the moderate future.  When it occurs whoever is making the change now needs to look at this test and understand it and the intent.  They've got to figure out "ok, why was it so important that there were exactly four kinds of stimuli here" and it'll take them a few hours to learn, I think, that actually this isn't critical at all, but that instead we're just looking to assert that there are only a few stimuli.

Is my thinking here correct?  We don't actually care that the number is 4, we just care that it's something like "probably more than 1, but also way less than the length of the story".  If that's true, then let's assert **that**, rather than assert a very specific value here.

We shouldn't assert things that we don't really care about.  I don't think that we really care that this number is four.

Reply:
> A change of that sort is likely to occur in the moderate future. When it occurs whoever is making the change now needs to look at this test and understand it and the intent. They've got to figure out "ok, why was it so important that there were exactly four kinds of stimuli here" and it'll take them a few hours to learn, I think, that actually this isn't critical at all, but that instead we're just looking to assert that there are only a few stimuli.

> Is my thinking here correct? We don't actually care that the number is 4, we just care that it's something like "probably more than 1, but also way less than the length of the story". If that's true, then let's assert that, rather than assert a very specific value here.

> We shouldn't assert things that we don't really care about. I don't think that we really care that this number is four.

I've thought about this further after watching the testing philosophy video. I personally do care about changes of this sort and want to know about them, even at the cost of spending time understanding why they occurred. And I would want other developers to be aware that their changes modify things because the distributed cluster is a complex stateful system. IMO there are places for testing via interfaces, but not at the lowest level.

`assert len(stimulus_ids) < len(story) / 3` for example provokes a strong code smell response in me. It might protect a developer from  having to understand a failed test, but IMO it makes it *more* difficult to understand what's being tested and more importantly if there is a problem it might:

1. Hide it from the developer which I think it is a big deal in the context of distributed stability and deadlocks.
2. Incur even more cognitive overhead in understanding the abstraction that has been built into the test and whether it's wrong/right.

At least with an explicit test I know what the expectation was before the change and can update expectations going forward. I personally think that given the current focus on stability, the tradeoff between

1. a higher dev velocity and,
2. developers having to spend time developing a deeper understanding of distributed

would lead me to choose (2)

Reply:
This was historically my choice as well.  After doing this for a while though I found that modest changes to the codebase required me to change hundreds of different tests.  It was not a productive use of time, and only I was capable of changing the codebase.  As a result of this experience we now prefer tests that try to avoid having developers to learn about unrelated parts of the codebase.

If we all have to learn about the entire codebase then I don't think that we'll be able to move at all.

I'd be happy to engage further in this conversation, but for the moment I'm -1 on this PR for as long as it's testing in this way.  I think that the inertia that it adds to the codebase does not outweigh the benefit of the feature.

Reply:
Phrasing this differently, the only reason you're able to write this PR without touching several thousand of lines of code and spend a few months studying is due to this policy.

Reply:
`len(stimulus_ids) < len(story) / 3` is a property that should be true all the time but it is not the most interesting property for what we're looking at here.

These IDs we are asserting on are incredibly high level. 
- "compute something"
- "something is done"
- "retry something"
- "something is done, again"

If I had a reasonable way of encoding this information without hardcoding the actual names (like update-graph-hlg), I would prefer writing the explicit four events out. I considered using the uniqueness was the compromise we agreed on.

If we want to relax this, we can set it to `assert len(stimulus_ids) >= 4` but I would be perfectly fine keeping it the way it is. Anybody who is doing a change to this system _intentionally_ will not require more than a few seconds debugging this if it would ever fail and we gain

Reply:
Unresolving this conversation